File: platform/bukkit/src/main/java/top/wcpe/wcpelib/bukkit/command/entity/Command.java
package top.wcpe.wcpelib.bukkit.command.entity;

import org.bukkit.command.CommandSender;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import top.wcpe.wcpelib.bukkit.command.BukkitCommand;
import top.wcpe.wcpelib.bukkit.command.intel.ExecuteComponentFunctional;
import top.wcpe.wcpelib.bukkit.command.intel.TabCompleterFunctional;
import top.wcpe.wcpelib.common.command.v2.ChildCommand;
import top.wcpe.wcpelib.common.command.v2.ParentCommand;
import top.wcpe.wcpelib.common.command.v2.SingleCommand;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * 子命令类
 *
 * @author WCPE
 * @date 2021年4月23日 下午4:47:57
 * @deprecated This class is deprecated. Use the {@link SingleCommand} Or {@link ParentCommand} And {@link ChildCommand}.
 */
@Deprecated
public class Command implements BukkitCommand {
    /**
     * 名称
     */
    private final String name;
    /**
     * 参数
     */
    private final List<top.wcpe.wcpelib.common.command.CommandArgument> args = new ArrayList<>();
    /**
     * 是否隐藏无权限帮助
     */
    private final boolean hideNoPermissionHelp;
    /**
     * 介绍
     */
    private final String describe;
    /**
     * 权限
     */
    private final String permission;
    /**
     * 无权限提示
     */
    private final String noPermissionMessage;
    /**
     * 是否只能玩家用
     */
    private final boolean onlyPlayerUse;
    /**
     * 不是玩家使用提示
     */
    private final String noPlayerMessage;
    /**
     * 命令执行组件
     */
    private final top.wcpe.wcpelib.common.command.CommandExecute executeComponent;
    /**
     * tab补全器
     */
    private final TabCompleterFunctional tabCompleter;


    private Command(Command.Builder builder) {
        this.name = builder.name;
        this.hideNoPermissionHelp = builder.hideNoPermissionHelp;
        this.args.addAll(builder.args);
        this.describe = builder.describe;
        this.permission = builder.permission;
        this.noPermissionMessage = builder.noPermissionMessage;
        this.onlyPlayerUse = builder.onlyPlayerUse;
        this.noPlayerMessage = builder.noPlayerMessage;
        this.executeComponent = (sender, args) -> builder.executeComponent.execute((CommandSender) sender.getAdapter(), args);
        this.tabCompleter = builder.tabCompleter;
    }

    @Nullable
    @Override
    public TabCompleterFunctional getTabCompleter() {
        return tabCompleter;
    }

    @NotNull
    @Override
    public String getName() {
        return name;
    }

    @NotNull
    @Override
    public String getDescribe() {
        return describe;
    }

    @NotNull
    @Override
    public List<top.wcpe.wcpelib.common.command.CommandArgument> getArgs() {
        return args;
    }

    @Override
    public boolean getHideNoPermissionHelp() {
        return hideNoPermissionHelp;
    }

    @Nullable
    @Override
    public String getPermission() {
        return permission;
    }

    @NotNull
    @Override
    public String getNoPermissionMessage() {
        return noPermissionMessage;
    }

    @Override
    public boolean getOnlyPlayerUse() {
        return onlyPlayerUse;
    }

    @NotNull
    @Override
    public String getNoPlayerMessage() {
        return noPlayerMessage;
    }

    @Nullable
    @Override
    public top.wcpe.wcpelib.common.command.CommandExecute getExecuteComponent() {
        return executeComponent;
    }


    public static class Builder {
        private final String name;
        private final String describe;
        private final List<CommandArgument> args = new ArrayList<>();
        private boolean hideNoPermissionHelp = false;
        private String permission;
        private String noPermissionMessage = "§c你莫得 §6%permission% §c权限!";
        private boolean onlyPlayerUse = false;
        private String noPlayerMessage = "§c该指令只能玩家使用!";
        private ExecuteComponentFunctional executeComponent;
        private TabCompleterFunctional tabCompleter;

        public Builder(String name, String describe) {
            this.name = name;
            this.describe = describe;
        }

        public Builder args(CommandArgument... args) {
            Collections.addAll(this.args, args);
            return this;
        }

        public Builder hideNoPermissionHelp(boolean hideNoPermissionHelp) {
            this.hideNoPermissionHelp = hideNoPermissionHelp;
            return this;
        }

        public Builder permission(String permission) {
            this.permission = permission;
            return this;
        }

        public Builder noPermissionMessage(String noPermissionMessage) {
            this.noPermissionMessage = noPermissionMessage;
            return this;
        }

        public Builder onlyPlayerUse(boolean onlyPlayerUse) {
            this.onlyPlayerUse = onlyPlayerUse;
            return this;
        }

        public Builder noPlayerMessage(String noPlayerMessage) {
            this.noPlayerMessage = noPlayerMessage;
            return this;
        }

        public Builder executeComponent(ExecuteComponentFunctional executeComponent) {
            this.executeComponent = executeComponent;
            return this;
        }

        public Builder tabCompleter(TabCompleterFunctional tabCompleter) {
            this.tabCompleter = tabCompleter;
            return this;
        }

        public Command build() {
            return new Command(this);
        }
    }
}

File: platform/bungeecord/src/main/kotlin/top/wcpe/wcpelib/bungeecord/command/Command.kt
package top.wcpe.wcpelib.bungeecord.command

import top.wcpe.wcpelib.common.command.Command
import top.wcpe.wcpelib.common.command.CommandArgument
import top.wcpe.wcpelib.common.command.CommandExecute

/**
 * 由 WCPE 在 2022/9/12 13:58 创建
 *
 * Created by WCPE on 2022/9/12 13:58
 *
 * GitHub  : https://github.com/wcpe
 * QQ      : 1837019522
 * @author : WCPE
 * @since  : v1.1.0-alpha-dev-4
 */
@Deprecated("replace common command v2")
class Command(
    override val name: String,
    override val describe: String,
    override val args: List<CommandArgument>,
    override val hideNoPermissionHelp: Boolean,
    override val permission: String?,
    override val noPermissionMessage: String,
    override val onlyPlayerUse: Boolean,
    override val noPlayerMessage: String,
    override val executeComponent: CommandExecute?,
    override val tabCompleter: TabCompleter?
) : BungeeCordCommand {

    class Builder(private val name: String, private val describe: String) {
        private val args = mutableListOf<CommandArgument>()
        private var hideNoPermissionHelp = false
        private var permission: String? = null
        private var noPermissionMessage: String = "§c你莫得 §6%permission% §c权限!"
        private var onlyPlayerUse: Boolean = false
        private var noPlayerMessage: String = "§c该指令只能玩家使用!"
        private var executeComponent: CommandExecute? = null
        private var tabCompleter: TabCompleter? = null


        fun args(vararg args: CommandArgument): Builder {
            this.args.addAll(args)
            return this
        }

        fun hideNoPermissionHelp(hideNoPermissionHelp: Boolean): Builder {
            this.hideNoPermissionHelp = hideNoPermissionHelp
            return this
        }

        fun permission(permission: String?): Builder {
            this.permission = permission
            return this
        }

        fun noPermissionMessage(noPermissionMessage: String?): Builder {
            this.noPermissionMessage = noPermissionMessage!!
            return this
        }

        fun onlyPlayerUse(onlyPlayerUse: Boolean): Builder {
            this.onlyPlayerUse = onlyPlayerUse
            return this
        }

        fun noPlayerMessage(noPlayerMessage: String?): Builder {
            this.noPlayerMessage = noPlayerMessage!!
            return this
        }

        fun executeComponent(executeComponent: CommandExecute): Builder {
            this.executeComponent = executeComponent
            return this
        }

        fun tabCompleter(tabCompleter: TabCompleter): Builder {
            this.tabCompleter = tabCompleter
            return this
        }

        fun build(): Command {
            return Command(
                name,
                describe,
                args,
                hideNoPermissionHelp,
                permission,
                noPermissionMessage,
                onlyPlayerUse,
                noPlayerMessage,
                executeComponent,
                tabCompleter
            )
        }
    }
}